### ISSUE TITLE
Extra space with tab completion in ipython

### ISSUE THREAD

**Issue by isofer** [ContinuumIO/anaconda-issues#42]
This issue still persists.
see details: https://groups.google.com/a/continuum.io/d/topic/anaconda/oRV1Yc-H3IY/discussion

**Comment by orodbhen** [comment]
I've read over this thread at least a couple times, and I not sure if anyone found a solution or even a reliable workaround. Could someone who understands the issue well please provide a summary of what's causing it? For my benefit, and the benefit of anyone else who reads this in the future.

Is it the readline package? Is it the built-in python module? Is it ncurses? Is it the way Continuum compiles readline, prior to compiling Python?

We've been using a custom-built Python 2.7 distribution that I created on our CentOS 6 machines, and I haven't seen this problem there. Neither have I seen it in the Python packages installed on Fedora. Installing ipython with pip in a virtualenv doesn't have this issue either. So I'm a little confused about what makes the anaconda version different.

**Comment by orodbhen** [comment]
Ok. I'm going to give it a try. First this, from @kbg:

"The problem with the Anaconda Python package is visible inside the pyconfig.h file that comes with Anaconda, where the HAVE_LIBREADLINE and all the HAVE_RL_* entries are not defined. Apparently the Python readline module was built anyway, but without all the code inside the HAVE_RL_* ifdefs, which already handles the trailing spaces."

So the python readline module is compiled with a bad pyconfig.h. Why is it that way?

What I don't understand is why there were problems compiling Python 2.7.6 against the system readline library. I was able to compile 2.7.6 on CentOS 6, and readline.so was created without problems. I used:

    LDFLAGS="-Wl,-rpath,/opt/python/intel64/lib" ./configure --enable-unicode=ucs4 --enable-shared --prefix=/opt/python/intel64 --with-valgrind

Is it because you were using the anaconda library path? Is it a conflict between a library in that path, and the system libreadline?

So then what's needed, If understand correctly, is for Continuum to fix the bad pyconfig.h, and package ncurses with Anaconda to satisfy the readline.so dependency. Correct?

**Comment by immerrr** [comment]
> Is it a conflict between a library in that path, and the system libreadline?

There is a conflict between system ncurses & anaconda one: common linux distros build separate ncurses & libtinfo libs (with a special ncurses build flag) while conda ships a single ncurses binary. Python build script detects system-wide libtinfo.so and makes readline.so dependent on it, but since the library is not installed in the env, it is not found and this breaks readline.so at runtime.

There are two fixes: the better one, from @kbg, patches python build script so that ncurses/tinfo autodetection runs on conda-supplied library, and mine, which mitigates the issue patching ncurses build script to build separate libtinfo library so that it is installed into the env and resolved properly.

Alas, Continuum guys seem to disregard the issue.

**Comment by immerrr** [comment]
@orodbhen , addressing your other question

> Are these kinds of issues typical with Anaconda, or is it normally quite stable?

Conda employs some non-100%-reliable techniques, like patching strings that look like paths in the binaries. Most of packages are thoroughly tested, so they are quite stable as long as you don't do something unusual. At least they worked quite well for me.

However, there are corner cases. For example I have run into a case when such binary patching corrupted debug info and made valgrind choke.

Another issue I have run into is that conda-shipped qt library is built without ssl support (https://github.com/ContinuumIO/anaconda-issues/issues/140). Even though I'd love to help, it is far from trivial to find out what prevents qt from using openssl package without seeing the qt recipe, and I'm not sure if the [publicly available recipe](https://github.com/conda/conda-recipes/tree/master/qt) is compatible with the one Continuum uses for building packages in official channel.

**Comment by orodbhen** [comment]
Thanks @immerrr . It makes a lot more sense now. I have a Qt 4 build in my custom Python install with SSL enabled, that works quite well. But I'm not very familiar with the conda build system.

I was actually in the process of creating a tool for managing python packages and virtual envs, as well as non-python packages. I was looking at making a python clone of homebrew for Linux, with some changes. But then I discovered that conda does all of the above. Even if Continuum's packages prove unreliable, it at least provides a management framework that may work for us. 

 I had a heck of a time getting Python to build 32 bit on a 64 bit system. It's build scripts don't do a great job at honoring CFLAGS and LDFLAGS.

Anyways, I'll try the patch and see if it solves the problem for me.

Why does "conda search ncurses" not return any results?

**Comment by orodbhen** [comment]
The patch from @kbg fixed the issue for me.

It also fixed a segfault I was getting with a Traitsui script, when trying to run it in the Anaconda env. I wonder what other differences there are in Continuum's build recipe for Python.

**Comment by orodbhen** [comment]
How do I keep conda from downgrading Python to the original version every time I install something? Do I have to give the patched Python the same version number as the original?

**Comment by ccordoba12** [comment]
I'll try to review and merge the mentioned patch this week

**Comment by ccordoba12** [comment]
Ok, so I went through the entire thread trying to understand its main points. I have one question for @kbg:

The configure step in your Python recipe is

```bash
./configure --prefix=$PREFIX \
    --enable-shared --enable-ipv6 --enable-unicode=ucs4 \
    --with-tcltk-includes="-I$PREFIX/include" \
    --with-tcltk-libs="-L$PREFIX/lib -ltcl8.5 -ltk8.5" \
    CPPFLAGS="-I$PREFIX/include" \
    LDFLAGS="-L$PREFIX/lib -Wl,-rpath=$PREFIX/lib,--no-as-needed"
```

How important are the tcltk lines? I mean, are they really necessary?

**Comment by ccordoba12** [comment]
Just as a side note for @immerrr: the error you mentioned about Qt is solved now, as almost all other important issues related to it :-)

**Comment by ccordoba12** [comment]
I tested @kbg suggestion and everything seems to be working as expected!!

I submitted a PR to our internal repo, so let's hope this is fixed soon :-)

**Comment by immerrr** [comment]
@ccordoba12 that is so awesome, thank you!

**Comment by ccordoba12** [comment]
Ok guys, we just released a new version of Python 2 (specifically `2.7.9-3`) which finally fixes this problem.

Sorry for the big delay in addressing this issue, and thanks a lot to @kbg for providing a very informed and clear solution.

Happy coding!! :-)

**Comment by immerrr** [comment]
@ccordoba12, will python3 be updated soon too?

**Comment by jjhelmus** [comment]
Thanks for fixing this! I can verify the new build, 2.7.9-3, no longer inserts an extra tab.

**Comment by ccordoba12** [comment]
@immerrr, I thought Python 3 doesn't have this problem. I'll verify and report back :-)

**Comment by ccordoba12** [comment]
Ok, I tested it in Python 3 and @immerrr is right, that's a problem there too.

We'll try to release a new package soon to fix this problem :-)

**Comment by jjhelmus** [comment]
The extra space issue seems to have been fixed in Python 3 with the release of python 3.4.3-1 and python 3.3.5-4.  Thanks!

**Comment by ccordoba12** [comment]
@jjhelmus, thanks for testing!! So this infamous bug can finally be closed , yay!! \o/
